Implement useOnyx hook

fabioh8010 opened the pull request:
### Details

This PR adds a new way to consume Onyx data in the application, the `useOnyx` hook.

Relevant links:

- [Slack proposal](https://expensify.slack.com/archives/C01GTK53T8Q/p1706718802348819)
- [Proposal doc](https://docs.google.com/document/d/1oUDZH73gaeSM_8A5wuG1hc6L1-8c3XmVHfxLXxjN7YY/edit)

### Related Issues

https://github.com/Expensify/App/issues/34339

### Automated Tests

N/A

### Manual Tests

N/A

### Author Checklist

- [x] I linked the correct issue in the `### Related Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->

</details>

fabioh8010 commented on lib/useOnyx.ts:
```diff
@@ -0,0 +1,132 @@
+
+const cache = new Map<string, unknown>();
+
+function useOnyxWithSyncExternalStore<TKey extends OnyxKey>(key: TKey, options?: UseOnyxOptions<TKey>): OnyxValue<TKey> {
```
This is just for testing purposes @roryabraham , I will keep only the `useSyncExternalStore` version before opening the PR 😄

blazejkustra commented on .eslintrc.js:
```diff
@@ -1,5 +1,6 @@
 module.exports = {
-    extends: ['expensify', 'prettier'],
+    extends: ['expensify', 'prettier', 'plugin:react-hooks/recommended'],
```
[Done](https://github.com/Expensify/eslint-config-expensify/pull/85) ✅ 

But this resulted in a cascade of updates in the config, not sure if we want to proceed.

blazejkustra replied:
@fabioh8010 Let's update eslint-config-expensify in this PR and remove 'plugin:react-hooks/recommended' here

blazejkustra commented on lib/useOnyx.ts:
```diff
@@ -0,0 +1,144 @@
+     * **While the store has not changed, repeated calls to getSnapshot must return the same value.**
+     * If the store changes and the returned value is different (as compared by Object.is), React re-renders the component.
+     *
+     * When the `key` is changed (e.g. to get a different record from a collection) and it's not yet in the cache,
+     * we return the value from the previous key to avoid briefly returning a `null` value to the component, thus avoiding a useless re-render.
```
I raised a concern in a DM that this may led to some unexpected bugs in NewDot. Fabio's argument was that this is to omit unnecessary re-renders, but with a status property when a key is changed it will be re-rendered either way. 

I'm not convinced that this logic is needed.

fabioh8010 replied:
I've removed after my last changes (wasn't being able to reproduce it anymore)

fabioh8010 commented on lib/useOnyx.ts:
```diff
@@ -0,0 +1,200 @@
+    /**
+     * TODO: Check if we still need this flag and associated logic.
+     */
+    // allowStaleData?: boolean;
```
@roryabraham @tgolen Could you explain me what this `allowStaleData` property do exactly in `withOnyx` HOC?

tgolen replied:
I'll defer most of my knowledge about it to this code comment:

https://github.com/Expensify/react-native-onyx/blob/10b74b7fd8b68ba9280b14015d62d1c6cefe3026/lib/withOnyx.js#L66-L76

tgolen replied:
My basic understanding is that it allows the component to be rendered immediately instead of the next tick due to cache optimization.

hannojg replied:
Yeah I think this is still needed.

- A component can render immediately if the value for a certain onyx key has already been cached
- However, the component we want to render might misses to be rendered immediately if the value is currently being updated
- This flag enables to render the component with "stale" data (the data before the update will happen)

(Reading through that code makes me wonder why we can't update the cache directly once we merge new values, but thats probably a different topic)

hannojg replied:
One side note: I still have to review the whole PR, maybe the usage of that prop become obsolete (let me double check 🔍 )

fabioh8010 replied:
@hannojg Please share if you have some basic example about how to test this flag with `withOnyx` (we have only a few usages in the app), so I can also compare here if the behavior is consistent with `useOnyx`.

hannojg replied:
Push okay, so actually after reviewing the PR I feel like maybe this exact option is needed anymore actually!

However, I am wondering, what happens to the use case of `!Onyx.hasPendingMergeForKey(key)`:

https://github.com/Expensify/react-native-onyx/blob/10b74b7fd8b68ba9280b14015d62d1c6cefe3026/lib/withOnyx.js#L77-L79

It has been initially added here:

- https://github.com/Expensify/react-native-onyx/pull/303

I believe right now we are missing that behaviour and the `useOnyx` hook would show stale data, e.g. meaning this behaviour:

```js
Onyx.merge('report_123', newValue); 
Navigation.navigate('report_screen', 'report_123');

// In ReportScreen

const data = useOnyx('report_123')
```

data would first be `staleData` and then in a second render become `newValue`.

Right now we are explicitly not using the cache value if has `Onyx.hasPendingMergeForKey(key)`.
However we made it possible to allow for reading stale data by that flag. 

So I think both should be added back:

- The  `Onyx.hasPendingMergeForKey(key)` for reading from cache or nah
- The `allowStaleData` flag

fabioh8010 replied:
Thanks for the example @hannojg ❤️ I will try to reproduce here and do the adjustments if necessary

blazejkustra commented on lib/useOnyx.ts:
```diff
@@ -0,0 +1,200 @@
+function useOnyx<TKey extends OnyxKey, TReturnData = OnyxValue<TKey>>(key: TKey, options?: UseOnyxOptions<TKey, TReturnData>): UseOnyxData<TReturnData> {
+    const connectionIDRef = useRef<number | null>(null);
+    const previousKey = usePrevious(key);
+    const previousDataRef = useRef<TReturnData | null>(null);
```
Why not just `dataRef`? OR `currentDataRef`?
```suggestion
    const currentDataRef = useRef<TReturnData | null>(null);
```

blazejkustra commented on lib/useOnyx.ts:
```diff
@@ -0,0 +1,200 @@
+         *
+         * If they are equal, we just return the previous internal data.
+         */
+        const newData = getCachedValue(key, options?.selector);
+        if (!deepEqual(previousDataRef.current, newData)) {
+            previousDataRef.current = newData as TReturnData;
+        }
+
+        return previousDataRef.current as TReturnData;
```
This is the same code as above in Case 2, right?

fabioh8010 replied:
Is almost the same, in Case 2 we have a condition. I kept them separated to explain better the possible cases.

blazejkustra commented on lib/useOnyx.ts:
```diff
@@ -0,0 +1,200 @@
+                     * We don't need to update the Onyx cache again here, when `callback` is called the cache is already
+                     * expected to be updated, so we just signal that the store changed and `getSnapshot()` can be called.
+                     */
+                    fetchStatusRef.current = 'loaded';
+                    onStoreChange();
```
Does it matter if we swap these?
```suggestion
                    onStoreChange();
                    fetchStatusRef.current = 'loaded';
```

fabioh8010 commented on lib/useOnyx.ts:
```diff
@@ -0,0 +1,206 @@
+    return (Onyx.tryGetCachedValue(key, {selector}) ?? null) as OnyxValue<TKey>;
+}
+
+function useOnyx<TKey extends OnyxKey, TReturnData = OnyxValue<TKey>>(key: TKey, options?: UseOnyxOptions<TKey, TReturnData>): UseOnyxData<TReturnData> {
```
Currently there is a minor type problem here when using the selector, I'm working on a fix.

blazejkustra commented on package.json:
```diff
@@ -58,11 +58,12 @@
     "eslint-plugin-import": "^2.29.1",
     "eslint-plugin-jsx-a11y": "^6.8.0",
     "eslint-plugin-react": "^7.31.10",
+    "eslint-plugin-react-hooks": "^4.6.0",
```
You can remove this plugin now, please double check `eslint-plugin-react-hooks` from eeslint-config-expensify works 🫣

fabioh8010 replied:
Seems to be working 🙂

blazejkustra commented:
Lint is failing, you can either fix these errors (add {} to each new Promise), or turn of the rule

hannojg commented on .eslintrc.js:
```diff
@@ -67,6 +67,7 @@ module.exports = {
                         tsx: 'never',
                     },
                 ],
+                'rulesdir/prefer-onyx-connect-in-libs': 'off',
```
[DUMB QUESTION] Why exactly did we turn that off here?

fabioh8010 replied:
This rule is complaining in my TS files because I'm calling `Onyx` methods

fabioh8010 replied:
@tgolen Actually I just saw that [we are suppressing the rule in withOnyx.js too](https://github.com/Expensify/react-native-onyx/blob/main/lib/withOnyx.js#L288). I don't think is a problem with the ESLint, just that it doesn't make too much sense to have this rule enabled in this repository since `rulesdir/prefer-onyx-connect-in-libs` looks like tailored for E/App. For example, this is the error that is shown:

> Only call Onyx.connect() from inside a /src/libs/** file. React components and non-library code should not use Onyx.connect()

It doesn't apply to here.

tgolen replied:
OK, thanks! I guess it might make more sense if the rule was turned off by default and then only enabled in the App repo, since it's very specific to that repo. I'll have to keep this in mind for future custom ESLint rules.

hannojg commented on lib/useOnyx.ts:
```diff
@@ -0,0 +1,208 @@
+    status: FetchStatus;
+};
+
+function isCollectionMemberKey<TKey extends OnyxKey>(key: TKey): boolean {
```
Can we reuse `isCollectionMemberKey` from Onyx.js here?

fabioh8010 replied:
I'm not sure, I saw that we already have a `isCollectionMemberKey` in `Onyx.js` but it actually expects two params (`collectionKey` and `key`), while I have only `key` in the hook. I will check again if we can reuse it somehow

fabioh8010 replied:
Actually I think we can reuse it, will do it

hannojg reviewed (commented):
Code looks 👌👌👌 to me, just wondering if we should add some unit tests for the hook?

[Need to check one more use case in detail, will update in a moment]

hannojg reviewed (changes requested):
After more careful consideration I think we need to be double careful about this one use case to not introduce a regression

roryabraham commented on lib/useOnyx.ts:
```diff
@@ -0,0 +1,189 @@
+        }
+
+        return currentDataRef.current as CachedValue<TKey, TReturnData>;
+    }, [key, options?.selector]);
```
It's worth noting that if the selector or options are unstable references, then `getSnapshot` will be redefined with each render.

We could consider using `useRef` to stabilize the reference internally so that caller's don't have to be wary of this pitfall:

```diff
diff --git a/lib/useOnyx.ts b/lib/useOnyx.ts
index 965cf4f..9b454f9 100644
--- a/lib/useOnyx.ts
+++ b/lib/useOnyx.ts
@@ -78,6 +78,9 @@ function useOnyx<TKey extends OnyxKey, TReturnData = OnyxValue<TKey>>(key: TKey,
         );
     }, [previousKey, key]);
 
+    const selector = useRef(options?.selector);
+    selector.current = options?.selector;
+
     /**
      * According to React docs, `getSnapshot` is a function that returns a snapshot of the data in the store that’s needed by the component.
      * **While the store has not changed, repeated calls to getSnapshot must return the same value.**
@@ -89,7 +92,7 @@ function useOnyx<TKey extends OnyxKey, TReturnData = OnyxValue<TKey>>(key: TKey,
          *
          * We just return the data from the Onyx cache.
          */
-        if (!Onyx.isCollectionKey(key) && !options?.selector) {
+        if (!Onyx.isCollectionKey(key) && !selector.current) {
             return getCachedValue(key);
         }
 
@@ -103,13 +106,13 @@ function useOnyx<TKey extends OnyxKey, TReturnData = OnyxValue<TKey>>(key: TKey,
          *
          * If they are equal, we just return the previous internal data.
          */
-        const newData = getCachedValue(key, options?.selector);
+        const newData = getCachedValue(key, selector.current);
         if (!deepEqual(currentDataRef.current, newData)) {
             currentDataRef.current = newData as CachedValue<TKey, TReturnData>;
         }
 
         return currentDataRef.current as CachedValue<TKey, TReturnData>;
-    }, [key, options?.selector]);
+    }, [key, selector]);
 
     const subscribe = useCallback(
         (onStoreChange: () => void) => {
```

I've also seen some people wrap this pattern with a simple `useLiveRef` utility hook:

```ts
function useLiveRef<T>(value: T) {
  const ref = useRef(value);
  ref.current = value;
  return ref;
}

...

const selector = useLiveRef(options?.selector);
```

That would allow us to safely use this hook like so:

```ts
const report = useOnyx(ONYXKEYS.COLLECTION.REPORT, {
    selector: (reports) => reports[reportID],
});
```

Instead of having to do some verbose memoization:

```ts
const selector = useCallback((reports) => reports[reportID], [reportID]);
const options = useMemo(() => ({selector}), [selector]);
const report = useOnyx(ONYXKEYS.COLLECTION.REPORT, options);
```

fabioh8010 replied:
That makes sense to me, thanks for the suggestion!

roryabraham commented on lib/useOnyx.ts:
```diff
@@ -0,0 +1,189 @@
+    const subscribe = useCallback(
+        (onStoreChange: () => void) => {
+            connectionIDRef.current = Onyx.connect({
+                key: key as CollectionKeyBase,
```
Why would this be a `CollectionKeyBase`? Couldn't it be any `OnyxKey` _or_ `CollectionKeyBase`?

fabioh8010 replied:
It's because I'm supplying a boolean to `waitForCollectionCallback` and that is making TS require a `CollectionKeyBase` here, but as this connect usage is generic it's fine to assert to `CollectionKeyBase` imo.

roryabraham reviewed (changes requested):
Also, let's add some automated UI tests to cover this hook? There are some examples in `withOnyxTest` we can follow.

Otherwise this is looking pretty good. It's so simple compared with `withOnyx` 🙂

fabioh8010 commented on lib/useOnyx.ts:
```diff
@@ -0,0 +1,189 @@
+        }
+
+        /**
+         * Case 2 - We have a non-collection/collection key with/without selector
```
I will adjust to `Case 2 - We have a non-collection/collection key and/or selector` since Case 2 also handles non-collection keys, wdyt?

roryabraham replied:
collection key and/or selector expands to:

"collection key and selector or selector"

Which is logically equivalent to:

"collection key and selector or (not collection key and selector)"

whatever you think is clearest though is fine. It's just a comment

fabioh8010 commented:
Most of the comments addressed, the only pending is the `allowStaleData` flag and the tests which I will continue tomorrow.

fabioh8010 commented on lib/Onyx.d.ts:
```diff
@@ -115,6 +120,21 @@ declare const METHOD: {
+ * @param key - The collection member key to split.
+ * @returns A tuple where the first element is the collection part and the second element is the ID part.
+ */
+declare function splitCollectionMemberKey<TKey extends CollectionKey>(key: TKey): [TKey extends `${infer Prefix}_${string}` ? `${Prefix}_` : never, string];
```
@blazejkustra Maybe you have some idea how to improve these typings? It should allow only collection member keys, but currently is allowing collection keys too.

```ts
Onyx.splitCollectionMemberKey(`${ONYXKEYS.COLLECTION.POLICY}${policyID}`); // ["policy_", string], correct
Onyx.splitCollectionMemberKey(ONYXKEYS.COLLECTION.POLICY); // ["policy_", string], should raise an error in the param
```

blazejkustra replied:
How about this?

```ts
type CollectionKeyBase = 'policy_' | 'report_';
type RemoveUnderscore<T> = T extends `${infer Word}_` ? Word : never;

const ONYXKEYS = {COLLECTION: {POLICY: 'policy_'}} as const;

declare function splitCollectionMemberKey<TKey extends RemoveUnderscore<CollectionKeyBase>, TRest extends string>(key: TRest extends '' ? never : `${TKey}_${TRest}`): [TKey, TRest];

splitCollectionMemberKey(`${ONYXKEYS.COLLECTION.POLICY}${123}`); // ["policy_", string], correct
splitCollectionMemberKey(ONYXKEYS.COLLECTION.POLICY);
```

blazejkustra replied:
I'm not sure if it's worth it, as this type is quite complicated, up to you 👍

fabioh8010 replied:
Yeah, let's avoid overcomplicating and skip this for now.

tgolen commented on lib/Onyx.js:
```diff
@@ -213,6 +212,15 @@ function isCollectionMemberKey(collectionKey, key) {
+ * @returns {[String, String]} A tuple where the first element is the collection part and the second element is the ID part.
+ */
+function splitCollectionMemberKey(key) {
+    return key.split('_');
```
Should this include `isCollectionMemberKey(key) && key.split('_')` so that it's a little bit more safe (say someone tries to use an `_` in the name of a key that's not a collection)?

tgolen commented on lib/useOnyx.ts:
```diff
@@ -0,0 +1,193 @@
+    // allowStaleData?: boolean;
+
+    /**
+     * Sets an initial value to be returned by the hook during the first render.
```
```suggestion
     * This value will be returned by the hook on the first render while the data is being read from Onyx.
```

tgolen commented on lib/useOnyx.ts:
```diff
@@ -0,0 +1,193 @@
+    initialValue?: OnyxValue<TKey>;
+
+    /**
+     * If included, this will be used to subscribe to a subset of an Onyx key's data.
```
```suggestion
     * This will be used to subscribe to a subset of an Onyx key's data.
```

fabioh8010 commented on lib/useOnyx.ts:
```diff
@@ -0,0 +1,193 @@
+        const previousCollectionKey = Onyx.splitCollectionMemberKey(previousKey)[0];
+        const collectionKey = Onyx.splitCollectionMemberKey(key)[0];
+
+        if (Onyx.isCollectionMemberKey(previousCollectionKey, previousKey) && Onyx.isCollectionMemberKey(collectionKey, key) && previousCollectionKey === collectionKey) {
```
These conditions serve to ensure we can only handle dynamic collection member keys from the same collection, for example:

- From `account` to `session`: ❌ 
- From `policy_` to `report_`: ❌ 
- From `report_` to `report_1283`: ❌ 
- From `report_385` to `report_9385`: ✅ 

Basically we have only one case where we would supply a different key to `useOnyx` over the time, when we are dealing with collection member keys:

```ts
const {value: policy} = useOnyx(`${ONYXKEYS.COLLECTION.POLICY}${policyID}`);
```

This way we ensure that `useOnyx` subscription is dealing with only one type of data and guarantee stable typings and safeness to the consumer.

fabioh8010 commented:
I've addressed review comments from @tgolen and implemented `allowStaleData` logic cc @hannojg . Currently I'm investigating an issue I'm facing with the loading states, which are not being updated correctly when mounting a component that uses `useOnyx`.

roryabraham commented on lib/useOnyx.ts:
```diff
@@ -0,0 +1,171 @@
+
+type CachedValue<TKey extends OnyxKey, TValue> = TValue extends OnyxValue<TKey> ? TValue : TKey extends CollectionKeyBase ? NonNullable<OnyxCollection<TValue>> : TValue;
+
+type UseOnyxData<TKey extends OnyxKey, TValue> = {
```
Hmmm, I thought we had aligned on this as a better pattern for `useOnyx` return data that's the best of both worlds:

- simple to get the "main" return value
- easy to extend with more configs / extra fields

```suggestion
type UseOnyxData<TKey extends OnyxKey, TValue> = [CachedValue<TKey, TValue>, {status: FetchStatus}];
```

that way we can do something like this:

```ts
const [account] = useOnyx(ONYXKEYS.ACCOUNT)l
const [reports, {status: reportsLoadingStatus}] = useOnyx(ONYXKEYS.COLLECTION.REPORT);
```

roryabraham commented:
Also, would love to see some automated UI tests added, similar to what we have for `withOnyx`

fabioh8010 commented:
Hi @tgolen I'm currently fixing the loading state problems, after that will proceed with the tests.

fabioh8010 commented:
@roryabraham @tgolen @blazejkustra I've pushed the changes to re-write implementation fix the loading states (and other issues I've found during tests). Key changes:
- Change `getSnapshot()` to return the entire hook data (`[CachedValue<TKey, TValue>, {status: FetchStatus}]`) in a stable way so it prevents issues with wrong loading state.
- Moved `initialValue` and `allowStaleData` logic to inside `getSnapshot()` function in order to consolidate all logic and fix several issues with stale data.

fabioh8010 commented:
@roryabraham @tgolen @blazejkustra I've added some unit tests for `useOnyx`, I decided to use `renderHook` approach to test the hook and focus more on its result behavior during the tests, please let me know what do you think.

I didn't implement unit tests for `allowStaleData`, not sure yet how to implement them properly and I also noticed that we don't have it for `withOnyx` too.

blazejkustra commented on lib/useOnyx.ts:
```diff
@@ -0,0 +1,188 @@
+import useLiveRef from './useLiveRef';
+import usePrevious from './usePrevious';
+
+type OnyxValue<TKey extends OnyxKey> = TKey extends CollectionKeyBase ? OnyxCollection<KeyValueMapping[TKey]> : OnyxEntry<KeyValueMapping[TKey]>;
```
Let's move this type next to OnyxKey

blazejkustra commented on lib/useOnyx.ts:
```diff
@@ -0,0 +1,188 @@
+
+type UseOnyxData<TKey extends OnyxKey, TValue> = [
+    CachedValue<TKey, TValue>,
+    {
+        status: FetchStatus;
+    },
```
How about extracting it to a separate type?

blazejkustra commented on lib/useOnyx.ts:
```diff
@@ -0,0 +1,188 @@
+
+        // If the previously cached data is different from the new data, we update both cached data
+        // and the result data to be returned by the hook.
+        // We can't directly compare the value from `dataRef` with `newData` because we default `undefined`
+        // to `null` when setting `dataRef` value to ensure we have a consistent return, but we need to be able to differentiate
+        // between `undefined` and `null` during the comparison, so `cachedValueRef` is used to store the real value without changing
+        // to `null`.
```
This part seems quite complicated, had to read through this couple of times before I understood what is going on 😅 

One question, why can't we use null value for the loading status as well? This way we could simplify the logic a bit here

fabioh8010 replied:
Let's just simplify then, wdyt?

```ts
        // If the previously cached data is different from the new data, we update both cached data
        // and the result data to be returned by the hook.
        if (!deepEqual(cachedValueRef.current, newData)) {
            cachedValueRef.current = newData as CachedValue<TKey, TReturnData>;

            // If the new data is `undefined` we default it to `null` to ensure the consumer get a consistent result from the hook.
            dataRef.current = [(cachedValueRef.current ?? null) as CachedValue<TKey, TReturnData>, {status: newFetchStatus ?? 'loaded'}];
        }
```

> One question, why can't we use null value for the loading status as well? This way we could simplify the logic a bit here

I didn't understand quite well the question 😅 The comments on that condition was more directed to the use and difference between `cachedValueRef` and `dataRef`, the `newFetchStatus ?? 'loaded'` part I'm just defaulting to `loaded` in case it was not set before.

blazejkustra replied:
Sorry! I meant why do we need to use undefined here:
```ts
        // If we have pending merge operations for the key during the first connection, we set the new value to `undefined`
        // and fetch status to `loading` to simulate that it is still being loaded until we have the most updated data.
        // If `allowStaleData` is `true` this logic will be ignored and cached value will be used, even if it's stale data.
        if (isFirstConnectionRef.current && Onyx.hasPendingMergeForKey(key) && !options?.allowStaleData) {
            newValue = undefined;
            newFetchStatus = 'loading';
        }
```

Most likely I'm missing something but if we were to use null in the logic above we could compare `dataRef` with `newData`, no?

roryabraham commented on lib/useOnyx.ts:
```diff
@@ -0,0 +1,189 @@
+
+    // Stores the previous cached value as it's necessary to compare with the new value in `getSnapshot()`.
+    // We initialize it to `undefined` to simulate that we don't have any value from cache yet.
+    const cachedValueRef = useRef<CachedValue<TKey, TReturnValue> | undefined>(undefined);
```
*question:* since this is used to store the previous cached value, why don't we use `usePrevious` for this?

roryabraham replied:
seems like you could do this:

```ts
const prevCachedValue = usePrevious(resultRef.current[0]);
```

and that might be a bit more concise

roryabraham replied:
Not sure I got all that but I'll take your word for it 😅

roryabraham commented on lib/useOnyx.ts:
```diff
@@ -0,0 +1,189 @@
+    selector?: Selector<TKey, unknown, TReturnValue>;
+};
+
+type FetchStatus = 'loading' | 'loaded';
```
let's use an object / const here and export it so we can check loading status with a const in E/App?

roryabraham replied:
sounds good

roryabraham commented on tests/unit/useOnyxTest.ts:
```diff
@@ -0,0 +1,384 @@
+        it('should return value and loaded state when loading cached key', async () => {
+            Onyx.set(ONYXKEYS.TEST_KEY, 'test');
+
+            const {result} = renderHook(() => useOnyx(ONYXKEYS.TEST_KEY));
```
`renderHook` seems like a great utility for this! nice TIL

roryabraham reviewed (changes requested):
tests look great! everything looks great! I seriously considered merging this but I think the last blocker for me is that we should use a constant for loading state 🙂

blazejkustra reviewed (approved):
LGTM! 🚀 

I tested it in the App today, also migrated some withOnyx usage to check if it works as expected, looks very promising! Great work @fabioh8010

tgolen commented:
Does there need to be anything generated for the API docs for this?

fabioh8010 commented:
@tgolen I've ran the script to update the docs

roryabraham commented:
LGTM 🚀

mountiny commented:
As noted by others, the code changes are looking great! Not leaving a review so I don't take credits for weeks worth of work, but wanted to note that you have done great job with this project, not only Fabio and Blazej, but also Rory and Tim (and others leaving a review on the doc) for pushing this forwards, thank you! 🏅
